technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+* Create GrafanaDataSource object
+* Create Prometheus object
+* End reconciliation loop
+
+#### Event: CassandraDatacenter ready
+
+##### cassandra cluster controller
+
+* Update status to reflect state of CassandraCluster
+* End reconciliation loop
+
+##### reaper controller
+
+* Start reconciliation loop after requeue
+* Create Job to initialize reaper keyspace
+* Requeue request until job finishes
+
+##### stargate controller
+
+* Start reconciliation loop after requeue
+* Create stargate Deployment
+* Requeue request until Deployment ready
+
+#### Event: CassandraCluster status updated
+
+##### k8ssandra controller
+
+* Start reconciliation loop
+  * Update status to reflect state of CassandraCluster
+* End reconciliation loop
+
+#### Event: Reaper schema job completed
+
+##### reaper controller
+
+* Create Deployment
+* Requeue request until Deployment ready
+* End reconciliation loop
+
+#### Event: Stargate ready
+
+##### k8ssandra controller
+
+* Update status to reflect state of Stargate
+
+#### Event: Reaper ready
+
+##### k8ssandra controller
+
+* Update status to reflect state of Reaper
+
+#### Event: Grafana and Prometheus ready
+
+##### monitoring controller
+
+* Update status of MonitoringConfiguration to reflect state of Grafana and Prometheus
+
+#### Event: MonitoringConfiguration status updated
+
+##### k8ssandra controller
+
+* Update status to reflect state of MonitoringConfiguration
+
+#### Notes
+
+The k8ssandra controller creates several objects:
+
+* CassandraCluster
+* Reaper
+* Stargate
+* MonitoringConfiguration
+
+Whenever any of these objects is updated, it should trigger a reconciliation of the k8ssandra controller. In the reconciliation loop, the status of each object will essentially be pushed up into the K8ssandraCluster object.
+
+### Change Cassandra heap settings
+
+The user updates the heap settings in the K8ssandraCluster manifest and applies the changes with kubectl apply.
+
+#### Event: K8ssandraCluster updated
+
+##### k8ssandra controller
+
+* Update CassandraCluster 
+* End reconciliation loop
+
+#### Event: CassandraCluster updated
+
+##### cassandra cluster controller
+
+* Update CassandraDatacenter(s)
+  * Update CassandraCluster status
+* Requeue reconciliation request
+
+#### Event: CassandraCluster status updated
+
+##### k8ssandra controller
+
+* Update status to reflect state of CassandraCluster
+
+#### Event: CassandraDatacenter update complete and ready
+
+##### cassandra cluster controller
+
+* Update status of CassandraCluster
+* End reconciliation loop
+
+#### Event: CassandraCluster status updated
+
+##### k8ssandra controller
+
+* Update status to reflect state of CassandraCluster
+
+#### Notes
+
+Notice that the user does not directly update the CassandraDatacenter. The CassandraDatacenter is owned and managed by the CassandraCluster controller. Updates made directly to the CassandraDatacenter would potentially be lost. Changes need to be propagated through the K8ssandraCluster object.
```

Comment:
Does this still allow for template change detection for the deployment for the new updates?  Or, does that end up impacting ALL of the K8ssandraCluster "children" objects?

Reply:
Not sure I understand the question but maybe some background will help. Each controller watches for events involving a particular custom resource. Those events include create, update, delete. The controller's job is to make the desired state of the object (as declare in its spec) match the actual state of the object.